File: Linked List/Leetcode/Remove Linked List Elements.cpp
/*
Remove all elements from a linked list of integers that have value val.

Example:

Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) 
    {
        if(!head)
            return NULL;
        
        if(!head->next)
        {
            if(head->val == val)
            {
                delete head;
                head = NULL;
            }
            return head;
        }
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode *curr = head,*prev = dummy,*aux;
        while(curr)
        {
            aux = curr->next;
            if(curr->val == val)
            {
                delete curr;
                prev->next = aux;
            }
            else
                prev = curr;
            curr = aux;
        }
        head = dummy->next;
        delete dummy;
        return head;
    }
};

File: Recursion/Remove Linked List Elements.java
/*
Given the head of a linked list and an integer val, remove all the nodes of the linked list that has Node.val == val, and return the new head.
Example 1:
Input: head = [1,2,6,3,4,5,6], val = 6
Output: [1,2,3,4,5]
Example 2:

Input: head = [], val = 1
Output: []
Example 3:

Input: head = [7,7,7,7], val = 7
Output: []
*/

//Iterative
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if(head == null)
            return null;
        if(head.next == null)
            return head.val == val?null:head;
        
        ListNode dummy = new ListNode(-1);
        ListNode res = dummy;
        
        while(head != null) {
            if(head.val != val) {
                res.next = head;
                res = res.next;
            }
            head = head.next;
        }
        res.next = null;
        return dummy.next;
    }
}

//Recursive
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if(head == null)
            return null;
        if(head.next == null)
            return head.val == val?null:head;
        ListNode res = head;
        res.next = removeElements(head.next,val);
        if(head.val == val)
            return res.next;
        return res;
    }
}
